sound/soc/sof/sof-audio.h

```diff
@@ -30,6 +30,30 @@
 
 #define WIDGET_IS_DAI(id) ((id) == snd_soc_dapm_dai_in || (id) == snd_soc_dapm_dai_out)
 
+struct snd_sof_widget;
+struct snd_sof_route;
+
+struct topology_widget_ipc_ops {
+	int widget_type;
+
+	/* function used for filling up the IPC structure to set up the widget */
+	int (*ipc_update)(struct snd_soc_component *scomp, struct snd_sof_widget *swidget);
+
+	/* function used for freeing the IPC-specific data when unloading the widget */
+	void (*widget_unload)(struct snd_soc_component *scomp, struct snd_sof_widget *swidget);
```

Comment:
ipc_widget_free (or whatever is the name for !setup)?

Reply:
fixed